# HG changeset patch
# User Guillaume Flandin <guillaume.offline@gmail.com>
# Date 1540585656 14400
#      Fri Oct 26 16:27:36 2018 -0400
# Node ID 456523d3cb4a5003cbbf90c444a969a8ab0e896e
# Parent  3a66ae436d8a7410e7f5df902763e76bd395959c
savefig: new function

* scripts/plot/util/savefig.m: New file.
* scripts/plot/util/module.mk: Update.
* plot.txi: Document.

diff --git a/doc/interpreter/plot.txi b/doc/interpreter/plot.txi
--- a/doc/interpreter/plot.txi
+++ b/doc/interpreter/plot.txi
@@ -1103,16 +1103,18 @@ stages it is often better to save the gr
 associated information so that changes---to colors, axis limits, marker styles,
 etc.---can be made easily from within Octave.  The @code{hgsave}/@code{hgload}
 commands can be used to save and re-create a graphics object.
 
 @DOCSTRING(hgsave)
 
 @DOCSTRING(hgload)
 
+@DOCSTRING(savefig)
+
 @node Interacting with Plots
 @subsection Interacting with Plots
 
 The user can select points on a plot with the @code{ginput} function or
 select the position at which to place text on the plot with the
 @code{gtext} function using the mouse.
 
 @DOCSTRING(ginput)
diff --git a/scripts/plot/util/module.mk b/scripts/plot/util/module.mk
--- a/scripts/plot/util/module.mk
+++ b/scripts/plot/util/module.mk
@@ -73,16 +73,17 @@ GEN_FCN_FILES += $(%canon_reldir%_GEN_FC
   %reldir%/pan.m \
   %reldir%/print.m \
   %reldir%/printd.m \
   %reldir%/refresh.m \
   %reldir%/refreshdata.m \
   %reldir%/rotate.m \
   %reldir%/rotate3d.m \
   %reldir%/saveas.m \
+  %reldir%/savefig.m \
   %reldir%/shg.m \
   %reldir%/struct2hdl.m \
   %reldir%/subplot.m \
   %reldir%/zoom.m
 
 %canon_reldir%dir = $(fcnfiledir)/plot/util
 
 %canon_reldir%_DATA = \
diff --git a/scripts/plot/util/savefig.m b/scripts/plot/util/savefig.m
new file mode 100644
--- /dev/null
+++ b/scripts/plot/util/savefig.m
@@ -0,0 +1,94 @@
+## Copyright (C) 2018 Guillaume Flandin
+##
+## This file is part of Octave.
+##
+## Octave is free software: you can redistribute it and/or modify it
+## under the terms of the GNU General Public License as published by
+## the Free Software Foundation, either version 3 of the License, or
+## (at your option) any later version.
+## your option) any later version.
+##
+## Octave is distributed in the hope that it will be useful, but
+## WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+## GNU General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with Octave; see the filename COPYING.  If not, see
+## <https://www.gnu.org/licenses/>.
+
+## -*- texinfo -*-
+## @deftypefn  {} {} savefig ()
+## @deftypefnx {} {} savefig (@var{filename})
+## @deftypefnx {} {} savefig (@var{h}, @var{filename})
+## @deftypefnx {} {} savefig (@var{h})
+## @deftypefnx {} {} savefig (@var{h}, @var{filename}, @qcode{"compact"})
+## Save graphics handle @var{h} to file @var{filename}.
+##
+## If unspecified, @var{h} is set to the current figure.
+##
+## If unspecified, @var{filename} is set to @file{"Untitled.fig"}.  If
+## @var{filename} does not have an extension, a default extension @file{".fig"}
+## will be added.
+##
+## A third input argument will be accepted but ignored, for Matlab compatibility.
+##
+## @seealso{hgsave, hdl2struct, openfig}
+## @end deftypefn
+
+function savefig (varargin)
+
+  ## Default values for input arguments
+  h = [];
+  filename = "Untitled.fig";
+
+  ## Check input arguments
+  if (nargin == 1)
+    if (ishghandle (varargin{1}))
+      h = varargin{1};
+    elseif (ischar (varargin{1}))
+      filename = varargin{1};
+    else
+      error ("savefig: invalid input");
+    endif
+  elseif (nargin == 2 || nargin == 3)
+    if (! ishghandle (varargin{1}))
+      error ("savefig: invalid figure handle");
+    endif
+    h = varargin{1};
+    if (! ischar (varargin{2}))
+      error ("savefig: invalid filename");
+    endif
+    filename = varargin{2};
+    # Input "compact" ignored (Matlab compatibility)
+  elseif (nargin > 3)
+    print_usage ();
+  endif
+
+  ## Check figure handle input
+  if (isempty (h))
+    h = gcf ();
+  endif
+
+  ## Check filename extension
+  [~, ~, ext] = fileparts (filename);
+  if (isempty (ext))
+    filename = [filename ".fig"];
+  endif
+
+  ## Save file
+  hgsave (h, filename);
+
+endfunction
+
+%!test
+%! unwind_protect
+%!   h = figure ("Visible", "off");
+%!   ftmp = [tempname() ".fig"];
+%!   savefig (h, ftmp);
+%!   savefig (ftmp);
+%!   savefig (h, ftmp, "compact");
+%! unwind_protect_cleanup
+%!   close (h);
+%!   unlink (ftmp);
+%! end_unwind_protect
